x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,342 @@
+        builder.endObject();
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        // We cannot fully rewrite the query to a NestedQueryBuilder here because that query builder validates that the path is
+        // registered as a nested object. This is not the case for semantic_text fields; the SemanticTextInferenceResultFieldMapper
+        // metafield mapper indexes inference results for the field using a "shadow" nested field mapping that is coordinated between
+        // this query and the mapper. This "shadow" mapping is not registered with the ES index mappings.
+        //
+        // Instead, we extract the inference results from the supplier in a serializable format and handle creation of the Lucene query
+        // in this class in doToQuery.
+        if (inferenceResults != null) {
+            return this;
+        }
+
+        if (inferenceResultsSupplier != null) {
+            inferenceResults = inferenceResultsSupplier.get();
+            return this;
+        }
+
+        Map<String, Set<String>> modelsForFields = computeModelsForFields(queryRewriteContext.getIndexMetadataMap().values());
+        Set<String> modelsForField = modelsForFields.get(fieldName);
+        if (modelsForField == null || modelsForField.isEmpty()) {
+            throw new IllegalArgumentException("Field [" + fieldName + "] is not a semantic_text field type");
+        }
+
+        if (modelsForField.size() > 1) {
+            // TODO: Handle multi-index semantic queries
+            throw new IllegalArgumentException("Field [" + fieldName + "] has multiple models associated with it");
+        }
+
+        InferenceAction.Request inferenceRequest = new InferenceAction.Request(
+            TaskType.ANY,
+            modelsForField.iterator().next(),
+            List.of(query),
+            Map.of(),
+            InputType.SEARCH
+        );
+
+        SetOnce<InferenceServiceResults> inferenceResultsSupplier = new SetOnce<>();
+        queryRewriteContext.registerAsyncAction(
+            (client, listener) -> executeAsyncWithOrigin(
+                client,
+                ML_ORIGIN,
+                InferenceAction.INSTANCE,
+                inferenceRequest,
+                listener.delegateFailureAndWrap((l, inferenceResponse) -> {
+                    inferenceResultsSupplier.set(inferenceResponse.getResults());
+                    l.onResponse(null);
+                })
+            )
+        );
+
+        return new SemanticQueryBuilder(this, inferenceResultsSupplier);
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        if (inferenceResults == null) {
+            throw new IllegalStateException("Query builder must be rewritten first");
+        }
+
+        List<? extends InferenceResults> inferenceResultsList = inferenceResults.transformToCoordinationFormat();
+        if (inferenceResultsList.isEmpty()) {
```

Comment:
Should these conditions be managed by the `SemanticTextIFieldMapper`? I'd say the responsibility of this class is to rewrite the query and invoke the proper query method in the `FieldMapper` (I know it's a pending conversation) 😉